Mathlib/AlgebraicGeometry/PullbackCarrier.lean

```diff
@@ -0,0 +1,356 @@
+/-
+Copyright (c) 2024 Qi Ge, Christian Merten, Andrew Yang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Qi Ge, Christian Merten, Andrew Yang
+-/
+import Mathlib.Algebra.Category.Ring.LinearAlgebra
+import Mathlib.AlgebraicGeometry.ResidueField
+
+/-!
+# Underlying topological space of fibre product of schemes
+
+Let `f : X ⟶ S` and `g : Y ⟶ S` be morphisms of schemes. In this file we describe the underlying
+topological space of `pullback f g`, i.e. the fiber product `X ×[S] Y`.
+
+## Main results
+
+- `AlgebraicGeomerty.Scheme.Pullback.carrierEquiv`: The bijective correspondence between the points
+  of `X ×[S] Y` and pairs `(z, p)` of triples `z = (x, y, s)` with `f x = s = g y` and
+  prime ideals `q` of `κ(x) ⊗[κ(s)] κ(y)`.
+- `AlgebraicGeometry.Scheme.Pullback.exists_preimage`: For every triple `(x, y, s)` with
+  `f x = s = g y`, there exists `z : X ×[S] Y` lying above `x` and `y`.
+
+We also give the ranges of `pullback.fst`, `pullback.snd` and `pullback.map`.
+
+-/
+
+open CategoryTheory Limits TopologicalSpace LocalRing TensorProduct
+
+noncomputable section
+
+namespace AlgebraicGeometry.Scheme.Pullback
+
+universe u
+
+variable {X Y S : Scheme.{u}} (f : X ⟶ S) (g : Y ⟶ S)
+
+/-- A `Triplet` over `f : X ⟶ S` and `g : Y ⟶ S` is a triple of points `x : X`, `y : Y`,
+`s : S` such that `f x = s = f y`. -/
+structure Triplet where
+  /-- The point of `X`. -/
+  x : X
+  /-- The point of `Y`. -/
+  y : Y
+  /-- The point of `S` below `x` and `y`. -/
+  s : S
+  hx : f.base x = s
+  hy : g.base y = s
+
+variable {f g}
+
+namespace Triplet
+
+@[ext]
+protected lemma ext {t₁ t₂ : Triplet f g} (ex : t₁.x = t₂.x) (ey : t₁.y = t₂.y) : t₁ = t₂ := by
+  cases t₁; cases t₂; simp; aesop
+
+/-- Make a triplet from `x : X` and `y : Y` such that `f x = g y`. -/
+@[simps]
+def mk' (x : X) (y : Y) (h : f.base x = g.base y) : Triplet f g where
+  x := x
+  y := y
+  s := g.base y
+  hx := h
+  hy := rfl
+
+variable (T : Triplet f g)
+
+/-- Given `x : X` and `y : Y` such that `f x = s = g y`, this is `κ(x) ⊗[κ(s)] κ(y)`. -/
+def tensor : CommRingCat :=
```

Comment:
I would prefer that arguments of `def`s are given explicitly, not via variables, if they are not too long. It is much more readable. 

(i.e. I usually put the `variable` statement *after* the def that needs it.)